src/providers/proxmoxve/cloudconfig.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{
+    network::{self, DhcpSetting, NetworkRoute},
+    providers::MetadataProvider,
+};
+use anyhow::Result;
+use ipnetwork::IpNetwork;
+use openssh_keys::PublicKey;
+use pnet_base::MacAddr;
+use serde::Deserialize;
+use slog_scope::warn;
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{BufRead, BufReader},
+    net::{AddrParseError, IpAddr},
+    path::Path,
+    str::FromStr,
+};
+
+#[derive(Debug)]
+pub struct ProxmoxVECloudConfig {
+    pub meta_data: ProxmoxVECloudMetaData,
+    pub user_data: Option<ProxmoxVECloudUserData>,
+    pub vendor_data: ProxmoxVECloudVendorData,
+    pub network_config: ProxmoxVECloudNetworkConfig,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudMetaData {
+    #[serde(rename = "instance-id")]
+    pub instance_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudUserData {
+    pub hostname: String,
+    pub manage_etc_hosts: bool,
+    pub fqdn: String,
+    pub chpasswd: ProxmoxVECloudChpasswdConfig,
+    pub users: Vec<String>,
+    pub package_upgrade: bool,
+    #[serde(default)]
+    pub ssh_authorized_keys: Vec<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudChpasswdConfig {
+    pub expire: bool,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudVendorData {}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfig {
+    pub version: u32,
+    pub config: Vec<ProxmoxVECloudNetworkConfigEntry>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfigEntry {
+    #[serde(rename = "type")]
+    pub network_type: String,
+    pub name: Option<String>,
+    pub mac_address: Option<String>,
+    #[serde(default)]
+    pub address: Vec<String>,
+    #[serde(default)]
+    pub search: Vec<String>,
+    #[serde(default)]
+    pub subnets: Vec<ProxmoxVECloudNetworkConfigSubnet>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ProxmoxVECloudNetworkConfigSubnet {
+    #[serde(rename = "type")]
+    pub subnet_type: String,
+    pub address: Option<String>,
+    pub netmask: Option<String>,
+    pub gateway: Option<String>,
+}
+
+impl ProxmoxVECloudConfig {
+    pub fn try_new(path: &Path) -> Result<Self> {
+        let user_data_file = BufReader::new(File::open(path.join("user-data"))?);
+        let mut user_data = None;
+
+        if let Some(first_line) = user_data_file.lines().next() {
+            if let Ok(first_line) = first_line {
+                if first_line.starts_with("#cloud-config") {
+                    user_data = serde_yaml::from_reader(File::open(path.join("user-data"))?)?;
+                }
+            }
+        }
```

Comment:
Hmm, opening the file twice looks a bit clunky. Why not just read in the whole file, and then parse it as YAML only if it starts with `#cloud-config\n`?